crates/bpf-common/src/containers/mod.rs

```diff
@@ -0,0 +1,283 @@
+    #[serde(rename = "rootfsImageID")]
+    rootfs_image_id: String,
+    #[serde(rename = "rootfsImageName")]
+    rootfs_image_name: String,
+}
+
+/// JSON configuration of libpod (podman, cri-o) images.
+#[derive(Debug, Deserialize)]
+struct LibpodImageConfig {
+    id: String,
+    digest: String,
+}
+
+/// Database schema of libpod.
+#[derive(Queryable, Selectable)]
+#[diesel(table_name = schema::libpod_db_container_config)]
+#[diesel(check_for_backend(diesel::sqlite::Sqlite))]
+struct LibpodDBContainerConfig {
+    json: String,
+}
+
+/// Container information used in Pulsar alerts and rules.
+#[derive(Clone, Debug, PartialEq, Eq, Deserialize, Serialize, Validatron)]
+pub struct ContainerInfo {
+    pub id: String,
+    pub name: String,
+    pub image: String,
+    pub image_digest: String,
+}
+
+impl fmt::Display for ContainerInfo {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "{{ id: {}, name: {}, image: {}, image_digest: {} }}",
+            self.id, self.name, self.image, self.image_digest
+        )
+    }
+}
+
+impl ContainerInfo {
+    pub fn from_container_id(id: ContainerId) -> Result<Self, ContainerError> {
+        match id {
+            ContainerId::Docker(id) => {
+                let path = PathBuf::from(DOCKER_CONTAINERS_PATH)
+                    .join(&id)
+                    .join("config.v2.json");
+                let file = File::open(&path).map_err(|source| ContainerError::ReadFile {
+                    source,
+                    path: path.clone(),
+                })?;
+
+                let reader = BufReader::new(file);
+                let config: DockerConfig = serde_json::from_reader(reader)
+                    .map_err(|source| ContainerError::ParseConfigFile { source, path })?;
+
+                let name = config.name;
+                let name = if let Some(name) = name.strip_prefix('/') {
+                    name.to_owned()
+                } else {
+                    name
+                };
+                let image = config.config.image;
+                let image_digest = config.image_digest;
+
+                Ok(Self {
+                    id,
+                    name,
+                    image,
+                    image_digest,
+                })
+            }
+            ContainerId::Libpod(container_id) => {
+                let (db_config, db_path, image_store_path) = find_libpod_container(&container_id)?;
+                let config: LibpodConfig =
+                    serde_json::from_str(&db_config.json).map_err(|source| {
+                        ContainerError::ParseConfigDB {
+                            source,
+                            path: db_path,
+                        }
+                    })?;
+                let image_id = config.rootfs_image_id;
+
+                let image_store_file =
+                    File::open(&image_store_path).map_err(|source| ContainerError::ReadFile {
```

Comment:
@banditopazzo To be fair - I'm not entirely happy with the solution, but I couldn't come up with anything better.

This is reading the `images.json`, which contains the list of all images that podman ever pulled. And unfortunately, here I'm doing this for every container, which is of course bad (O(n) lookup for the image digest for every container...).

But unfortunately, I couldn't find any file which would contain an information about a single image. There is also no information about images in the sqlite database - I tried looking at all tables:

```
❯ sqlite3 ~/.local/share/containers/storage/db.sql
SQLite version 3.42.0 2023-05-16 12:36:15
Enter ".help" for usage hints.
sqlite> .tables
ContainerConfig       ContainerState        PodConfig
ContainerDependency   ContainerVolume       PodState
ContainerExecSession  DBConfig              VolumeConfig
ContainerExitCode     IDNamespace           VolumeState
```

and the only useful table is `ContainerConfig`, which we are using here, but it contains only `rootfsImageID` and `rootfsImageName` - there is no digest.

The only alternative which comes to my mind is some kind of caching, or even keeping the information about every podman image internally in Pulsar - we could try to read `images.json` only on the initial run and when someone is pulling the image - but the second part would be tricky and not so trivial to implement.